core/domain/classifier_services.py

```diff
@@ -126,7 +125,19 @@ def handle_non_retrainable_states(exploration, state_names,
 
     state_names_to_retrieve = []
     for current_state_name in state_names:
-        old_state_name = new_to_old_state_names[current_state_name]
+        old_state_name = current_state_name
+        if current_state_name in (
+                exp_versions_diff.old_to_new_state_names.values()):
+            # The structure of ExplorationVersionsDiff's
```

Comment:
Do a single pass outside the for loop to build the reverse mapping here (locally) and use that. Or better still add that as a method to the exp_versions_diff object.

Reply:
Done.